### PR TITLE
Working on tests for data checking and filtering

### PR DESCRIPTION
**Pull request by fastier-li**
Added tests, which trigger a small improvement in `AvroDataReader`.

### CODE REVIEW

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 1]
```diff
@@ -171,29 +179,32 @@ final class Driver(val sc: SparkContext, val params: GameParams, implicit val lo
...
-        .tap(df => require(df.count > 0, "Didn't find any data point with a weight >= 0. Aborting."))
-    } else {
-      data
+  protected[training] def checkData(data: Option[DataFrame]): Option[DataFrame] =
+
+    data.map { dataframe =>
+      if (params.checkData) {
+        val bad = dataframe
+          .map { row => if (row.getAs[Double]("weight") <= 0.0) 1 else 0 }
+          .reduce { (a, b) => a + b }
```
Comment (joshvfleming) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: How about:

```scala
  .reduce(_ + _)
```

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 2]
```diff
@@ -171,29 +179,32 @@ final class Driver(val sc: SparkContext, val params: GameParams, implicit val lo
...
+        if (bad > 0)
+          throw new IllegalArgumentException(s"Found $bad data points with weights <= 0. Please fix data set.")
```
Comment (joshvfleming) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]:
```scala
  require(bad == 0, s"Found $bad data points with weights <= 0. Please fix data set.")
```

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 3]
```diff
@@ -245,13 +256,13 @@ final class Driver(val sc: SparkContext, val params: GameParams, implicit val lo
...
   protected[training] def prepareNormalizationContexts(
-      trainingData: DataFrame,
+      trainingData: Option[DataFrame],
       featureIndexMapLoaders: IndexMapLoaders,
       statistics: FeatureShardStatisticsOpt): Option[Map[FeatureShardId, NormalizationContext]] =
 
     Filter(params.normalizationType != NormalizationType.NONE) {
       statistics
-        .getOrElse(calculateStatistics(trainingData, featureIndexMapLoaders))
+        .getOrElse(calculateStatistics(trainingData.get, featureIndexMapLoaders))
```
Comment (joshvfleming) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: It's a little odd that this function accepts `Option` but assumes that there's a value. Maybe do this `get` before calling the function?

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 4]
```diff
@@ -222,13 +233,13 @@ final class Driver(val sc: SparkContext, val params: GameParams, implicit val lo
...
   protected[training] def calculateAndSaveFeatureShardStats(
-      trainingData: DataFrame,
+      trainingData: Option[DataFrame],
       featureIndexMapLoaders: IndexMapLoaders): FeatureShardStatisticsOpt =
 
     params.summarizationOutputDirOpt
       .map {
         (summarizationOutputDir: String) => {
-          calculateStatistics(trainingData, featureIndexMapLoaders)
+          calculateStatistics(trainingData.get, featureIndexMapLoaders)
```
Comment (joshvfleming) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Same thing here.

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala (hunk 2 above)
Comment (fastier-li) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Fixed.

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala (hunk 3 above)
Comment (fastier-li) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Fixed. Scala should have e.g.
`data List a = Empty | Cons a (List a) deriving (Show, Read, Eq, Ord)`

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala (hunk 4 above)
Comment (fastier-li) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Yeap, same thing.

File: photon-client/src/main/scala/com/linkedin/photon/ml/util/Utils.scala [hunk 5]
```diff
@@ -332,4 +332,12 @@ protected[ml] object Utils {
    * @return Some[T] if p or None
    */
   def Filter[T](p: => Boolean)(f: => T): Option[T] = if (p) Some(f) else None
+
+  def Chain[T](f1: => T)(p: => Boolean)(f2: T => T): T = {
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/util/Utils.scala]: This isn't being used anywhere - did you intend to use it in `Driver.run`?

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala [hunk 6]
```diff
@@ -447,6 +449,43 @@ class DriverTest extends SparkTestUtils with GameTestUtils with TestTemplateWith
...
+        "feature-name-and-term-set-path" -> featuresDir,
+        "feature-shard-id-to-feature-section-keys-map" -> "shard1:features",
+        "train-input-dirs" -> inputFile,
+        "fixed-effect-data-configurations" -> "global:shard1,2",
+        "fixed-effect-optimization-configurations" -> "global:100,1e-11,0.3,1,LBFGS,l2",
+        "updating-sequence" -> "global",
+        "validate-input-dirs" -> "",
+        "output-dir" -> s"$getTmpDir/output",
+        "num-iterations" -> "5",
+        "check-data" -> "true")
```
Comment (ashelkovnykov) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala]: You can probably replace most of these entries by reusing `fixedEffectToyRunArgs` or `fixedEffectSeriousRunArgs`

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala [hunk 7]
```diff
@@ -447,6 +449,43 @@ class DriverTest extends SparkTestUtils with GameTestUtils with TestTemplateWith
...
+      .listFiles
+      .filter(_.getName.endsWith(".avro"))
+      .map(f => Array(inputDir + "feature-lists", f.getAbsolutePath))
+  }
+
+  /**
+   * Make sure we filter out negative or zero sample weights, and throw an exception if there are no samples left after
+   * filtering.
+   */
+  @Test(dataProvider = "badWeightsInputs") //, expectedExceptions = Array(classOf[IllegalArgumentException]))
```
Comment (fastier-li) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala]: It's actually already the case. Uncommented the `expectedExceptions`.

File: photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala (hunk 6 above)
Comment (fastier-li) [inline photon-client/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala]: Actually not - the heart data derivative being used is somewhat different.

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 8]
```diff
@@ -58,41 +58,46 @@ final class Driver(val sc: SparkContext, val params: GameParams, implicit val lo
...
-    val featureIndexMapLoaders = Timed("prepare features") {
-      prepareFeatureMaps()
-    }
-    val trainingDataToVerify = Timed("read training data") {
-      readTrainingData(featureIndexMapLoaders)
+    /**
+     * Helper to avoid writing the same code for training and validation data: read some data set, and then optionally
+     * validate it, while timing each step separately.
+     */
+    def readAndCheck(reader: => Option[DataFrame], what: String): Option[DataFrame] = {
```
Comment (fastier-li) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Done.

File: photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala [hunk 9]
```diff
@@ -171,29 +179,32 @@ final class Driver(val sc: SparkContext, val params: GameParams, implicit val lo
...
-      data
-        .filter(col("weight") > 0.0)
-        .tap(df => require(df.count > 0, "Didn't find any data point with a weight >= 0. Aborting."))
-    } else {
-      data
+  protected[training] def checkData(data: Option[DataFrame]): Option[DataFrame] =
+
+    data.map { dataframe =>
+      if (params.checkData) {
+        val bad = dataframe
```
Comment (fastier-li) [inline photon-client/src/main/scala/com/linkedin/photon/ml/cli/game/training/Driver.scala]: Done.

File: photon-client/src/main/scala/com/linkedin/photon/ml/util/Utils.scala (hunk 5 above)
Comment (fastier-li) [inline photon-client/src/main/scala/com/linkedin/photon/ml/util/Utils.scala]: Thanks! Removed.

**Pull request merged**